x-pack/libbeat/management/managerV2.go

```diff
@@ -594,6 +596,12 @@ func (cm *BeatV2Manager) reload(units map[unitKey]*client.Unit) {
 
 	// reload the output configuration
 	if err := cm.reloadOutput(outputUnit); err != nil {
+		// The manager has already signalled the Beat to stop,
+		// there is nothing else to do. Trying to reload inputs
+		// will only lead to race conditions.
+		if errors.Is(err, errStoppingOnOutputChange) {
+			return
```

Comment:
This return bypasses several of the state updates at the end of this function. 

https://github.com/elastic/beats/blob/f75a2ed898bc883c451d6aab35cc60d17158f788/x-pack/libbeat/management/managerV2.go#L626-L655

Rather than return here it is probably better to have this check prevent us from executing reloadInputs specifically

https://github.com/elastic/beats/blob/f75a2ed898bc883c451d6aab35cc60d17158f788/x-pack/libbeat/management/managerV2.go#L622-L624

Also since this isn't really an error (it is an expected outcome of this function), you could just return a bool and an error where the bool indicates whether the beat will stop. 

That would prevent you from incorrectly setting this as a real error in https://github.com/elastic/beats/blob/f75a2ed898bc883c451d6aab35cc60d17158f788/x-pack/libbeat/management/managerV2.go#L632-L635

Reply:
> This return bypasses several of the state updates at the end of this function.

That was my intention because the rest of the function does not know the manager has already signalled a shutdown: 
https://github.com/belimawr/beats/blob/e3367eddcb6f03c3d7c38e9735999be1a9ca1cb9/x-pack/libbeat/management/managerV2.go#L697-L702

```go
	if cm.stopOnOutputReload && cm.lastOutputCfg != nil {
		cm.logger.Info("beat is restarting because output changed")
		_ = unit.UpdateState(client.UnitStateStopping, "Restarting", nil)
		cm.Stop()
		return errStoppingOnOutputChange
	}
```

That caused a race condition that would prevent Filebeat to be stopped by the manager (some inputs would never stop), but the connection to Elastic-Agent was already closed, hence the state updates from `BeatV2Manager.reload` would fail due to the closed connection while some inputs would still be running and trying to send data to the output. I only discovered that because my tests got flaky with a Filebeat process that would never exit.

Because the shutdown process is already signalled, and `BeatV2Manager.reload` does not know about it, any subsequent state updates are not quite correct. The state updates from https://github.com/elastic/beats/blob/f75a2ed898bc883c451d6aab35cc60d17158f788/x-pack/libbeat/management/managerV2.go#L626-L655 that you pointed out will be performed considering that the Beat will keep running, which I believe is not correct.

I don't see the point in trying to start/modify/stop an input concurrently with the shutdown process.

On my current implementation `errStoppingOnOutputChange` will never reach https://github.com/elastic/beats/blob/f75a2ed898bc883c451d6aab35cc60d17158f788/x-pack/libbeat/management/managerV2.go#L632-L635 because the method has already returned.

However I do agree that `errStoppingOnOutputChange` is technically not an error condition, so I'll update it to also return a bool value indicating that the shutdown has started.

Reply:
I see the point about the state updates happening concurrently with the shutdown process, so we aren't guaranteed the writes to the agent succeed.

I think this is worth a comment to prevent this question from being asked again.

Reply:
>I think this is worth a comment to prevent this question from being asked again.

I added it: https://github.com/belimawr/beats/blob/8c8bb9a219ebbcf90c3e63f20a784665ae86839e/x-pack/libbeat/management/managerV2.go#L598-L605